ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/computation/CommandStore.java

```diff
@@ -87,19 +95,31 @@ public void close() {
    * @param statementString The string of the statement to be distributed
    * @param statement The statement to be distributed
    * @param overwriteProperties Any command-specific Streams properties to use.
-   * @return The ID assigned to the statement
+   * @return The status of the enqueued command
    */
-  public CommandId distributeStatement(
+  @Override
+  public QueuedCommandStatus enqueueCommand(
       final String statementString,
       final Statement statement,
       final KsqlConfig ksqlConfig,
-      final Map<String, Object> overwriteProperties
-  ) throws KsqlException {
+      final Map<String, Object> overwriteProperties) {
     final CommandId commandId = commandIdAssigner.getCommandId(statement);
     final Command command = new Command(
         statementString,
         overwriteProperties,
         ksqlConfig.getAllConfigPropsWithSecretsObfuscated());
+    final QueuedCommandStatus status = new QueuedCommandStatus(commandId);
+    this.commandStatusMap.compute(
+        commandId,
+        (k, v) -> {
+          if (v == null) {
+            return status;
+          }
+          // We should fail registration if a future is already registered, to prevent
+          // a caller from receiving a future for a different statement.
+          throw new IllegalStateException("Concurrent command with id: " + commandId);
```

Comment:
Maybe a more descriptive message, like: `Another command with the same id (%s) is being executed.`